File: server/src/provider.ts
// Pull request providers: GitHub, GitLab, Gitea/Forgejo.
//
// Ported from `src-tauri/src/provider.rs` using Node's built-in fetch (the Rust
// backend used reqwest).

import { AppError, httpError } from "./lib/errors.ts";

export interface PullRequestRecord {
  id: string;
  number: number;
  title: string;
  author: string;
  state: string;
  sourceBranch: string;
  targetBranch: string;
  webUrl: string;
  updatedAt: string | null;
}

function text(value: unknown, pointer: string): string {
  let current: unknown = value;
  for (const segment of pointer.split("/").filter((segment) => segment.length > 0)) {
    if (typeof current !== "object" || current === null) return "";
    current = (current as Record<string, unknown>)[segment];
  }
  return typeof current === "string" ? current : "";
}

function number(value: unknown, pointer: string): number {
  let current: unknown = value;
  for (const segment of pointer.split("/").filter((segment) => segment.length > 0)) {
    if (typeof current !== "object" || current === null) return 0;
    current = (current as Record<string, unknown>)[segment];
  }
  return typeof current === "number" && Number.isInteger(current) ? current : 0;
}

function normalize(provider: string, value: Record<string, unknown>): PullRequestRecord {
  if (provider === "gitlab") {
    return {
      id: String(number(value, "/id")),
      number: number(value, "/iid"),
      title: text(value, "/title"),
      author: text(value, "/author/name"),
      state: text(value, "/state"),
      sourceBranch: text(value, "/source_branch"),
      targetBranch: text(value, "/target_branch"),
      webUrl: text(value, "/web_url"),
      updatedAt: (typeof value.updated_at === "string" && value.updated_at) || null,
    };
  }
  const html = text(value, "/html_url");
  return {
    id: text(value, "/id"),
    number: number(value, "/number"),
    title: text(value, "/title"),
    author: text(value, "/user/login"),
    state: text(value, "/state"),
    sourceBranch: text(value, "/head/ref"),
    targetBranch: text(value, "/base/ref"),
    webUrl: html.length > 0 ? html : text(value, "/url"),
    updatedAt: (typeof value.updated_at === "string" && value.updated_at) || null,
  };
}

export async function listPullRequests(
  provider: string,
  baseUrl: string,
  owner: string,
  repository: string,
  token: string,
): Promise<PullRequestRecord[]> {
  const normalized = provider.toLowerCase();
  const base = baseUrl.replace(/\/+$/, "");
  let url: string;
  const headers: Record<string, string> = {};

  switch (normalized) {
    case "github":
      url = `${base}/repos/${owner}/${repository}/pulls?state=all&per_page=100`;
      headers["Accept"] = "application/vnd.github+json";
      if (token) headers["Authorization"] = `Bearer ${token}`;
      break;
    case "gitlab":
      url = `${base}/projects/${encodeURIComponent(`${owner}/${repository}`)}/merge_requests?scope=all&per_page=100`;
      if (token) headers["PRIVATE-TOKEN"] = token;
      break;
    case "gitea":
    case "forgejo":
      url = `${base}/repos/${owner}/${repository}/pulls?state=all&limit=100`;
      if (token) headers["Authorization"] = `Bearer ${token}`;
      break;
    default:
      throw new AppError("invalid", `unsupported provider: ${provider}`);
  }

  let response: Response;
  try {
    response = await fetch(url, { headers });
  } catch (error) {
    throw httpError(`request to ${url} failed: ${(error as Error).message}`);
  }
  if (!response.ok) {
    throw httpError(`provider responded ${response.status} for ${url}`);
  }
  const body: unknown = await response.json().catch(() => null);
  if (!Array.isArray(body)) {
    throw new AppError("command", `provider response from ${url}: expected a JSON array`);
  }
  return body.map((value) =>
    normalize(normalized, (value ?? {}) as Record<string, unknown>),
  );
}

File: src-tauri/src/provider.rs
use serde_json::Value;
use crate::error::{AppError, AppResult};
use crate::models::PullRequestRecord;

pub async fn list_pull_requests(provider: &str, base_url: &str, owner: &str, repository: &str, token: &str) -> AppResult<Vec<PullRequestRecord>> {
    let client = reqwest::Client::new();
    let provider = provider.to_lowercase();
    let base = base_url.trim_end_matches('/');
    let (url, request) = match provider.as_str() {
        "github" => {
            let url = format!("{base}/repos/{owner}/{repository}/pulls?state=all&per_page=100");
            let request = client.get(&url).header("Accept", "application/vnd.github+json").bearer_auth(token);
            (url, request)
        }
        "gitlab" => {
            let project_path = format!("{owner}/{repository}");
            let project = urlencoding::encode(&project_path);
            let url = format!("{base}/projects/{project}/merge_requests?scope=all&per_page=100");
            let request = client.get(&url).header("PRIVATE-TOKEN", token);
            (url, request)
        }
        "gitea" | "forgejo" => {
            let url = format!("{base}/repos/{owner}/{repository}/pulls?state=all&limit=100");
            let request = client.get(&url).bearer_auth(token);
            (url, request)
        }
        _ => return Err(AppError::Invalid(format!("unsupported provider: {provider}"))),
    };
    let response = request.send().await?.error_for_status()?;
    let values: Vec<Value> = response.json().await?;
    values.into_iter().map(|value| normalize(&provider, value)).collect::<AppResult<Vec<_>>>()
        .map_err(|error| AppError::Command(format!("provider response from {url}: {error}")))
}
fn text(value: &Value, pointer: &str) -> String { value.pointer(pointer).and_then(Value::as_str).unwrap_or("").to_string() }
fn number(value: &Value, pointer: &str) -> u64 { value.pointer(pointer).and_then(Value::as_u64).unwrap_or(0) }
fn normalize(provider: &str, value: Value) -> AppResult<PullRequestRecord> {
    Ok(if provider == "gitlab" {
        PullRequestRecord { id: number(&value, "/id").to_string(), number: number(&value, "/iid"), title: text(&value, "/title"), author: text(&value, "/author/name"), state: text(&value, "/state"), source_branch: text(&value, "/source_branch"), target_branch: text(&value, "/target_branch"), web_url: text(&value, "/web_url"), updated_at: value.pointer("/updated_at").and_then(Value::as_str).map(str::to_string) }
    } else {
        PullRequestRecord { id: value.pointer("/id").map(ToString::to_string).unwrap_or_default(), number: number(&value, "/number"), title: text(&value, "/title"), author: text(&value, "/user/login"), state: text(&value, "/state"), source_branch: text(&value, "/head/ref"), target_branch: text(&value, "/base/ref"), web_url: { let html = text(&value, "/html_url"); if html.is_empty() { text(&value, "/url") } else { html } }, updated_at: value.pointer("/updated_at").and_then(Value::as_str).map(str::to_string) }
    })
}
